File: LV 2/스킬트리.py
from collections import deque

def solution(skill, skill_trees):
    answer = 0
    for sk in skill_trees:
        queue = deque(skill)
        flag = True
        for s in sk:
            if s in queue:
                if queue[0] == s: # 선행스킬이면 큐에서 빼줌
                    queue.popleft()
                else: # 선행스킬이 아니라면 잘못된 것이므로 false
                    flag = False
                    break
        
        if flag:
            answer += 1
                    
    return answer

File: LV 2/스킬트리.java
class Solution {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        for (String skillTree : skill_trees) {
            String convert = convert(skillTree, skill);
            if (skill.indexOf(convert) == 0) {
                answer += 1;
            }
        }
        return answer;
    }

    private String convert(String skillTree, String skill) {
        StringBuilder sb = new StringBuilder();

        for (char ch : skillTree.toCharArray()) {
            if (isContain(skill, ch)) {
                sb.append(ch);
            }
        }

        return sb.toString();
    }

    private boolean isContain(String skill, char ch) {
        return skill.chars()
                .filter(c -> c == (int)ch)
                .findAny()
                .isPresent();
    }
}
